src/main/java/com/linecorp/armeria/server/thrift/ThriftServiceCodec.java

```diff
@@ -82,13 +91,24 @@
      */
     private final Map<String, ThriftFunction> functions = new HashMap<>();
 
-    private final ThreadLocalTProtocol inProto = new ThreadLocalTProtocol();
-    private final ThreadLocalTProtocol outProto = new ThreadLocalTProtocol();
+    private static final Map<SerializationFormat, ThreadLocalTProtocol> FORMAT_TO_THREAD_LOCAL_IN_PROTOCOL =
+            createFormatToThreadLocalTProtocolMap();
+    private static final Map<SerializationFormat, ThreadLocalTProtocol> FORMAT_TO_THREAD_LOCAL_OUT_PROTOCOL =
+            createFormatToThreadLocalTProtocolMap();
 
-    ThriftServiceCodec(Object service, TProtocolFactory protoFactory) {
+    ThriftServiceCodec(Object service, SerializationFormat defaultSerializationFormat,
+                       Iterable<SerializationFormat> allowedSerializationFormats) {
+        requireNonNull(allowedSerializationFormats, "allowedSerializationFormats");
         this.service = requireNonNull(service, "service");
-        this.protoFactory = requireNonNull(protoFactory, "protoFactory");
-        serializationFormat = ThriftProtocolFactories.toSerializationFormat(protoFactory);
+        this.defaultSerializationFormat =
+                requireNonNull(defaultSerializationFormat, "defaultSerializationFormat");
+        EnumSet<SerializationFormat> allowedSerializationFormatsSet = EnumSet.noneOf(SerializationFormat.class);
+        allowedSerializationFormats.forEach(allowedSerializationFormatsSet::add);
+        this.allowedSerializationFormats = Collections.unmodifiableSet(allowedSerializationFormatsSet);
+        if (!this.allowedSerializationFormats.contains(defaultSerializationFormat)) {
+            throw new IllegalArgumentException(
+                    "allowedSerializationFormats must contain the defaultSerializationFormat.");
```

Comment:
Could we remove this check for user-friendliness? Otherwise a user would need to repeat the default format twice:

    new ThriftService(..., THRIFT_BINARY, THRIFT_BINARY)